lld/COFF/Writer.cpp

```diff
@@ -1421,8 +1421,19 @@ void Writer::assignAddresses() {
     // If /FUNCTIONPADMIN is used, functions are padded in order to create a
     // hotpatchable image.
     uint32_t padding = sec->isCodeSection() ? config->functionPadMin : 0;
+    MachineTypes prevECMachine = IMAGE_FILE_MACHINE_UNKNOWN;
 
     for (Chunk *c : sec->chunks) {
+      if (isArm64EC(ctx.config.machine) && sec->isCodeSection()) {
+        MachineTypes machine = c->getMachine();
+        if (machine != IMAGE_FILE_MACHINE_UNKNOWN) {
+          // We need additional alignment when crossing EC range baudaries.
+          if (prevECMachine != IMAGE_FILE_MACHINE_UNKNOWN &&
+              machine != prevECMachine)
+            virtualSize = alignTo(virtualSize, 4096);
+          prevECMachine = machine;
+        }
```

Comment:
I think that an early continue might simplify this:

```suggestion
         if (prevECMachine == machine || machine == IMAGE_FILE_MACHINE_UNKNOWN)
           continue;
         // We need additional alignment when crossing EC range boundaries.
         virtualSize = alignTo(virtualSize, 4096);
         prevECMachine = machine;
```

Reply:
We can't use continue here because we still need to run the rest of the loop. I looked as simplifying `IMAGE_FILE_MACHINE_UNKNOWN` handling, which inspired me to do more MSVC testing and finding a potentially interesting behavior. It seems to preserve the fact that a chunk is a data chunk even when it's merged to a code section. I need to think more about it and run more tests, we may want to do something similar.